src/util/codec/mysql/decimal.rs

```diff
@@ -11,770 +11,2008 @@
+        match self {
+            Res::Ok(t) => Res::Ok(f(t)),
+            Res::Truncated(t) => Res::Truncated(f(t)),
+            Res::Overflow(t) => Res::Overflow(f(t)),
+        }
+    }
+
+    pub fn unwrap(self) -> T {
+        match self {
+            Res::Ok(t) |
+            Res::Truncated(t) |
+            Res::Overflow(t) => t,
+        }
+    }
 
-macro_rules! enable_conv_for_int {
-    ($t:ty) => {
-        impl From<$t> for Decimal {
-            fn from(t: $t) -> Decimal {
-                Decimal {
-                    coeff: t.into(),
-                    exp: 0,
-                    fsp: 0,
-                }
-            }
+    pub fn is_ok(&self) -> bool {
+        match *self {
+            Res::Ok(_) => true,
+            _ => false,
         }
-    };
+    }
+
+    pub fn is_overflow(&self) -> bool {
+        match *self {
+            Res::Overflow(_) => true,
+            _ => false,
+        }
+    }
 }
 
-enable_conv_for_int!(u32);
-enable_conv_for_int!(u64);
-enable_conv_for_int!(u16);
-enable_conv_for_int!(u8);
-enable_conv_for_int!(i32);
-enable_conv_for_int!(i64);
-enable_conv_for_int!(i16);
-enable_conv_for_int!(i8);
-enable_conv_for_int!(usize);
-enable_conv_for_int!(isize);
+impl<T> Deref for Res<T> {
+    type Target = T;
 
-impl Decimal {
-    /// Convert a float number to decimal.
-    ///
-    /// This function will use float's canonical string representation
-    /// rather than the accurate value the float represent.
-    pub fn from_f64(f: f64) -> Result<Decimal> {
-        if !f.is_finite() {
-            return Err(invalid_type!("{} can't be convert to decimal'", f));
+    fn deref(&self) -> &T {
+        match *self {
+            Res::Ok(ref t) |
+            Res::Overflow(ref t) |
+            Res::Truncated(ref t) => t,
+        }
+    }
+}
+
+impl<T> DerefMut for Res<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        match *self {
+            Res::Ok(ref mut t) |
+            Res::Overflow(ref mut t) |
+            Res::Truncated(ref mut t) => t,
         }
+    }
+}
 
-        let s = format!("{}", f);
-        Decimal::from_str(&s)
+// A `Decimal` holds 9 words.
+const WORD_BUF_LEN: u8 = 9;
+// A word holds 9 digits.
+const DIGITS_PER_WORD: u8 = 9;
+// A word is 4 bytes i32.
+const WORD_SIZE: u8 = 4;
+const DIG_MASK: u32 = TEN_POW[8];
+const WORD_BASE: u32 = TEN_POW[9];
+const WORD_MAX: u32 = WORD_BASE - 1;
+const MAX_FRACTION: u8 = 30;
+const DIG_2_BYTES: &'static [u8] = &[0, 1, 1, 2, 2, 3, 3, 4, 4, 4];
+const FRAC_MAX: &'static [u32] = &[900000000, 990000000, 999000000, 999900000, 999990000,
+                                   999999000, 999999900, 999999990];
+
+macro_rules! word_cnt {
+    ($len:expr) => (word_cnt!($len, u8));
+    ($len:expr, $t:ty) => ({
+        (($len) + DIGITS_PER_WORD as $t - 1) / DIGITS_PER_WORD as $t
+    })
+}
+
+/// Return the first encoded decimal's length.
+pub fn dec_encoded_len(encoded: &[u8]) -> Result<usize> {
+    if encoded.len() < 2 {
```

Comment:
`dec_encoded_len` doesn't check if the decimal is correctly encoded. Hence 2 is ok.